pkg/bloomgateway/resolver.go

```diff
@@ -61,36 +62,48 @@ func (r *defaultBlockResolver) Resolve(ctx context.Context, tenant string, inter
 }
 
 func blocksMatchingSeries(metas []bloomshipper.Meta, interval bloomshipper.Interval, series []*logproto.GroupedChunkRefs) []blockWithSeries {
-	result := make([]blockWithSeries, 0, len(metas))
-
-	for _, meta := range metas {
-		for _, block := range meta.Blocks {
+	slices.SortFunc(series, func(a, b *logproto.GroupedChunkRefs) int { return int(a.Fingerprint - b.Fingerprint) })
 
-			// skip blocks that are not within time interval
-			if !interval.Overlaps(block.Interval()) {
-				continue
+	result := make([]blockWithSeries, 0, len(metas))
+	cache := make(map[bloomshipper.BlockRef]int)
+
+	// find the newest block for each series
+	for _, s := range series {
+		var b *bloomshipper.BlockRef
+		var ts time.Time
+
+		for i := range metas {
+			for j := range metas[i].Blocks {
+				block := metas[i].Blocks[j]
+				version := metas[i].Sources[j].TS
```

Comment:
Is this indexing by `j` correct? IIRC each meta will usually have only one Source, right? But `j` here is iterating the blocks.

Reply:
So, the number of sources is equal to the number of blocks.

Reply:
Fixed the incorrect population of the `Sources` in the builder (https://github.com/grafana/loki/pull/15481/commits/92692e21987ae19491bf47cd0f5f582adf0509f9) and adjusted the code on the querier (https://github.com/grafana/loki/pull/15481/commits/f566b92c5be4b6b5aecb802c52e04319b66f212f)